third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -114,61 +123,387 @@ lazy_static! {
         Mutex::new(lru::LruCache::new(VERIFIED_CACHE_SIZE));
 }
 
-pub(crate) struct StructNameCache {
-    data: RwLock<(
-        BTreeMap<StructIdentifier, StructNameIndex>,
-        Vec<StructIdentifier>,
-    )>,
+//
+// Loader
+//
+
+#[derive(Clone)]
+pub(crate) enum Loader {
+    V1(LoaderV1),
+    #[allow(dead_code)]
+    V2(LoaderV2<DummyVerifier>),
 }
 
-impl Clone for StructNameCache {
-    fn clone(&self) -> Self {
-        let inner = self.data.read();
-        Self {
-            data: RwLock::new((inner.0.clone(), inner.1.clone())),
+impl Loader {
+    pub(crate) fn new(natives: NativeFunctions, vm_config: VMConfig) -> Self {
+        Self::V1(LoaderV1 {
+            scripts: RwLock::new(ScriptCache::new()),
+            type_cache: RwLock::new(TypeCache::empty()),
+            struct_name_index_map: StructNameIndexMap::empty(),
+            natives,
+            invalidated: RwLock::new(false),
+            module_cache_hits: RwLock::new(BTreeSet::new()),
+            vm_config,
+        })
+    }
+
+    pub(crate) fn is_v1(&self) -> bool {
+        matches!(self, Self::V1(_))
+    }
+
+    pub(crate) fn vm_config(&self) -> &VMConfig {
```

Comment:
let's define a macro for this dispatches, and use them for all this boilerplate (might need to make names like ty_cache uniform though, or take a param to macro)

Reply:
Done